_posts/2025-09-23-quarkus-a2a-java-grpc.adoc

```diff
@@ -0,0 +1,346 @@
+
+Notice that the client's https://github.com/a2aproject/a2a-samples/tree/main/samples/java/agents/dice_agent_multi_transport/client/pom.xml[pom.xml] file contains dependencies on `io.github.a2asdk:a2a-java-sdk-client` and `io.github.a2asdk:a2a-java-sdk-client-transport-grpc`.
+
+The `a2a-java-sdk-client` dependency provides access to a `Client.builder` that we'll use to create our A2A client and also provides the ability for the client to support the JSON-RPC transport.
+
+The `a2a-java-sdk-client-transport-grpc` dependency provides the ability for the client to support the gRPC transport.
+
+Let's see how the `TestClient` uses the A2A Java SDK to create a `Client` that supports both gRPC and JSON-RPC:
+
+[source,java]
+----
+...
+// Fetch the public agent card
+AgentCard publicAgentCard = new A2ACardResolver(serverUrl).getAgentCard();
+
+// Create a CompletableFuture to handle async response
+final CompletableFuture<String> messageResponse = new CompletableFuture<>();
+
+// Create consumers for handling client events
+List<BiConsumer<ClientEvent, AgentCard>> consumers = getConsumers(messageResponse);
+
+// Create error handler for streaming errors
+Consumer<Throwable> streamingErrorHandler = (error) -> {
+    System.out.println("Streaming error occurred: " + error.getMessage());
+    error.printStackTrace();
+    messageResponse.completeExceptionally(error);
+};
+
+// Create channel factory for gRPC transport
+Function<String, Channel> channelFactory = agentUrl -> {
+    return ManagedChannelBuilder.forTarget(agentUrl).usePlaintext().build();
+};
+
+ClientConfig clientConfig = new ClientConfig.Builder()
+    .setAcceptedOutputModes(List.of("Text"))
+    .build();
+
+// Create the client with both JSON-RPC and gRPC transport support.
+// The A2A server agent's preferred transport is gRPC, since the client
+// also supports gRPC, this is the transport that will get used
+Client client = Client.builder(publicAgentCard) <1>
+    .addConsumers(consumers) <2>
+    .streamingErrorHandler(streamingErrorHandler) <3>
+    .withTransport(GrpcTransport.class, new GrpcTransportConfig(channelFactory)) <4>
+    .withTransport(JSONRPCTransport.class, new JSONRPCTransportConfig()) <5>
+    .clientConfig(clientConfig) <6>
+    .build();
+
+// Create and send the message
+Message message = A2A.toUserMessage(messageText);
+
+System.out.println("Sending message: " + messageText);
+client.sendMessage(message); <7>
+System.out.println("Message sent successfully. Waiting for response...");
+
+try {
+    // Wait for response with timeout
+    String responseText = messageResponse.get();
+    System.out.println("Final response: " + responseText);
+} catch (Exception e) {
+    System.err.println("Failed to get response: " + e.getMessage());
+    e.printStackTrace();
+}
+...
+----
+<1> We can use `Client.builder(publicAgentCard)` to create our A2A client. We need to pass in the `AgentCard`
+for the A2A server agent this client will be communicating with.
```

Comment:
Maybe instead of '...for the A2A server agent...' we could put '...retrieved from the A2A server agent...'